library-build-transformer/src/main/kotlin/io/realm/buildtransformer/RealmBuildTransformer.kt

```diff
@@ -39,123 +52,124 @@ val logger: Logger = LoggerFactory.getLogger("realm-build-logger")
+    private val input: ListProperty<RegularFile>,
+    private val output: RegularFileProperty,
+) {
+    init {
+        transform()
     }
 
-    override fun getReferencedScopes(): MutableSet<in QualifiedContent.Scope> {
-        return ImmutableSet.of()
-    }
-
-    override fun transform(context: Context?,
-                           inputs: MutableCollection<TransformInput>?,
-                           referencedInputs: MutableCollection<TransformInput>?,
-                           outputProvider: TransformOutputProvider?,
-                           isIncremental: Boolean) {
-        @Suppress("DEPRECATION")
-        super.transform(context, inputs, referencedInputs, outputProvider, isIncremental)
-
-        // Poor mans version of detecting variants, since the Gradle API does not allow us to
-        // register a transformer for only a single build variant
-        // https://issuetracker.google.com/issues/37072849
-        val transformClasses: Boolean = context?.variantName?.startsWith(flavorToStrip.toLowerCase()) == true
+    private fun transform() {
+        val outputProvider = JarOutputStream(
+            BufferedOutputStream(
+                FileOutputStream(
+                    output.get().asFile.absolutePath + ".tmp"
+                )
+            )
+        )
 
         val timer = Stopwatch()
         timer.start("Build Transform time")
-        if (isIncremental) {
-            runIncrementalTransform(inputs!!, outputProvider!!, transformClasses)
-        } else {
-            runFullTransform(inputs!!, outputProvider!!, transformClasses)
-        }
-        timer.stop()
-    }
 
-    private fun runFullTransform(inputs: MutableCollection<TransformInput>, outputProvider: TransformOutputProvider, transformClasses: Boolean) {
-        logger.debug("Run full transform")
-        val outputDir = outputProvider.getContentLocation("realmlibrarytransformer", outputTypes, scopes, Format.DIRECTORY)
-        val inputFiles: MutableSet<File> = mutableSetOf()
-        inputs.forEach {
-            it.directoryInputs.forEach {
-                // Non-incremental build: Include all files
-                val dirPath: String = it.file.absolutePath
-                it.file.walkTopDown()
-                        .filter { it.isFile }
-                        .filter { it.name.endsWith(".class") }
-                        .forEach { file ->
-                            file.packageHierarchyRootDir = dirPath
-                            file.shouldBeDeleted = transformClasses && specificFilesToStrip.find { file.absolutePath.endsWith(it) } != null
-                            inputFiles.add(file)
-                        }
+        val annotationDescriptor = createDescriptor(annotationQualifiedName.get())
+        val metadataCollector =
+            io.realm.buildtransformer.asm.visitors.AnnotationVisitor(annotationDescriptor)
+
+        forEachJarEntry { jarEntry, inputStream ->
```

Comment:
It is a bit unclear why we iterate the jar entries twice, some documentation for each block would be nice.

Reply:
Added some comments